### ISSUE TITLE
Question about constraints in LOBPCG Solver

### ISSUE THREAD

**Issue by sethwatts** [hypre-space/hypre#872]
I have a question about the constraints that can be applied to the eigenvector solutions in the LOBPCG eigensolver. Can these constraints be used to enforce positive definiteness on the system?

I am computing resonant modes of a linear elastic system and before essential boundary conditions are applied, the stiffness and mass matrices are symmetric positive semi-definite, each with null space dimension equal to the number of rigid body modes in the spatial dimension of the problem, i.e., three in two dimensions, and six in three dimensions. For some problems, an essential boundary makes sense, but in others it does not. Can I generate a null basis and use the basis vectors to constrain the solver? Will preconditioners such as BoomerAMG work in this case, or must the mass matrix be SPD on its own without the constraints?

Thank you.

**Comment by lobpcg** [comment]
Just ran into this question now.

No, the constraints in LOBPCG cannot be used to remove the null-space in both matrices, because the orthogonal complement to the constraints is made in the scalar product of the mass matrix that not definite.

You need to remove the null-space from all iterative multi-vectors, starting with the initial approximation. Multiply by the stiffness or mass matrices actually removes the null-space. You need to filter out the null-space from the output of the preconditioner. This cannot be done by using the constraints in LOBPCG as I explain above, but you can write the custom preconditioner that performs orthogonalization to the null-space on the top of a hyper preconditioner. I have not tested it but I think that should work.

I don't know how AMG would react if you just run it on a semi-define stiffness matrix as is. If it complains maybe try adding a small positive value on the diagonal to make it positive definite.

**Comment by sethwatts** [comment]
Thank you for your response! I should have closed the thread earlier when I found a solution. For the specific case I asked about here, I wrote a function to identify and constrain a minimal number of degrees of freedom in the system to suppress rigid body modes without otherwise biasing the resonant mode shapes.

For a related problem, however, this approach does not work, as the problem was singular, although with a known basis for the nullspace. In that case, doing as @lobpcg mentioned worked well. My implementation there used an MFEM Operator to perform the pre-multiplication that removes the null space from the vector X.
